sql/scan.go

```diff
@@ -147,11 +163,73 @@ func (n *scanNode) Err() error {
 	return n.err
 }
 
+func (n *scanNode) init() bool {
+	return n.initScan() && n.initExprs()
+}
+
+// initExprs initializes the render and filter expressions for the
+// scan. Initialization consists of replacing QualifiedName nodes with
+// ParenExpr nodes for which the wrapped expression can be changed for each
+// row.
+func (n *scanNode) initExprs() bool {
+	n.qvals = make(qvalMap)
+	for i := range n.render {
+		n.render[i], n.err = extractQVals(n.desc, n.qvals, n.render[i])
+		if n.err != nil {
+			return false
+		}
+	}
+	n.filter, n.err = extractQVals(n.desc, n.qvals, n.filter)
+	return n.err == nil
+}
+
+// initScan initializes (and performs) the key-value scan.
+//
+// TODO(pmattis): The key-value scan currently reads all of the key-value
+// pairs, but they could just as easily be read in chunks. Probably worthwhile
+// to separate out the retrieval of the key-value pairs into a separate
+// structure.
+func (n *scanNode) initScan() bool {
+	// Initialize our key/values.
+	if n.desc == nil {
+		// No table to read from, pretend there is a single empty row.
+		n.kvs = []client.KeyValue{}
+		n.primaryKey = []byte{}
+		return true
+	}
+	// Retrieve all of the keys that start with our index key prefix.
+	startKey := proto.Key(structured.MakeIndexKeyPrefix(n.desc.ID, n.desc.PrimaryIndex.ID))
+	endKey := startKey.PrefixEnd()
+	// TODO(pmattis): Currently we retrieve all of the key/value pairs for
+	// the table. We could enhance this code so that it retrieves the
+	// key/value pairs in chunks.
```

Comment:
Apparently I really want to make sure we address this. I'll remove this comment.